File: sdk/python/gantry/webhooks.py
"""Verify Gantry webhook signatures."""
from __future__ import annotations

import hmac
import hashlib


def verify_webhook_signature(body: bytes, header: str, secret: str) -> bool:
    """Verify X-Gantry-Signature header against the webhook secret."""
    expected = "sha256=" + hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, header)

File: sdk/typescript/src/webhooks.ts
import { createHmac, timingSafeEqual } from 'crypto';

/** Verify X-Gantry-Signature header against the webhook secret. */
export function verifyWebhookSignature(
  body: Buffer | string,
  header: string,
  secret: string,
): boolean {
  const payload = typeof body === 'string' ? Buffer.from(body) : body;
  const expected =
    'sha256=' + createHmac('sha256', secret).update(payload).digest('hex');
  try {
    return timingSafeEqual(Buffer.from(expected), Buffer.from(header));
  } catch {
    return false;
  }
}
